### PR TITLE
fix for issue #4087

### LINKED ISSUE #4087: Cannot add identity to PauliSum
**Description of the issue**

Identities cannot be added to PauliSums. This might be related to `cirq.IdentityGate` not being a `cirq.Pauli`.

**How to reproduce the issue**

```
ps = cirq.PauliSum()
ps += cirq.I(cirq.LineQubit(0))

# Generates the error:
# TypeError: unsupported operand type(s) for +=: 'PauliSum' and 'GateOperation'
```

**Cirq version**
0.11.0.dev

### PR DESCRIPTION
**Pull request by NoureldinYosri**
Problem: the PauliSum add function checks that the type of the other operand is one of numbers.Complex, PauliString, PauliSum and raises NotImplemented otherwise.
Fix: if the other operand is not one of the types try to cast it to PauliString and raise error only if that fails

Fixes #4087.

### CODE REVIEW

**Comment by NoureldinYosri** [comment]
Hi, I added a test, please let me know of any suggestions :)

**Comment by NoureldinYosri** [comment]
@balopat any updates ?

**Review by balopat — APPROVED** [review]

**Comment by balopat** [comment]
This looks good, but it seems that there are failing tests...

**Comment by NoureldinYosri** [comment]
@balopat 
looks like the only failing test is this one https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/circuits/circuit_test.py#L299 where in this line https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/circuits/circuit_test.py#L310 it expects the result of addition to be
```
cirq.Circuit([
     cirq.Moment(
         cirq.X(cirq.NamedQubit('a')),
     ),
 ],
)
```
but after fixing the addition operator it gives 
```
cirq.PauliSum(cirq.LinearDict({frozenset({(cirq.NamedQubit('a'), cirq.X)}): (1+0j), frozenset(): (1+0j)})
```

so before I dive deeper in debugging Is this the intended behaviour ?

**Comment by NoureldinYosri** [comment]
@balopat I restricted the casting to only happen for gate operations, this should fix the issue

**Comment by maffoo** [comment]
Ref: #4087

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 1]
```diff
@@ -1136,6 +1136,13 @@ def test_paulisum_validation():
     ld = cirq.LinearDict({key: 2.0})
     assert cirq.PauliSum(ld) == cirq.PauliSum.from_pauli_strings([2 * cirq.X(q[0])])
 
+    try:
+        ps = cirq.PauliSum()
+        ps += cirq.I(cirq.LineQubit(0))
+    except:
+        raise Exception("Failed to add identity to PauliSum.")
```
Comment (vtomole) [inline cirq-core/cirq/ops/linear_combinations_test.py:1143]: Remove this try/except

**Review by vtomole — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 1 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/linear_combinations_test.py]: done

**Comment by NoureldinYosri** [comment]
@vtomole @balopat a gentle reminder about this PR.

File: cirq-core/cirq/ops/linear_combinations.py [hunk 2]
```diff
@@ -570,6 +570,8 @@ def __iadd__(self, other):
         return self
 
     def __add__(self, other):
+        if isinstance(other, (gate_operation.GateOperation)):
```
Comment (balopat) [inline cirq-core/cirq/ops/linear_combinations.py:573]: As discussed on the Cirq Cynque, we'll need to make this more specific. 
Namely, the instance check should be about checking whether this is an instance of `identity.IdentityGate`. After this I think we're done with this PR :)

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/linear_combinations.py:573]: but that won't fix the issue in #4087 , the identity there is constructed as `cirq.I(cirq.LineQubit(0))` which has type `cirq.ops.gate_operation.GateOperation` not `cirq.ops.identity.IdentityGate`, so checking the instance type will fail.

Comment (balopat) [reply cirq-core/cirq/ops/linear_combinations.py:573]: Oh, yes, I meant instance GateOperation, other.gate is instance of IdentityGate.

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/linear_combinations.py]: updated (y).

**Pull request merged**